### PR TITLE
Add paragraph support to the table module

### PR DESCRIPTION
**Pull request by DokaRus**

### CODE REVIEW

File: formats/table/index.js [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+      const identityMarker = key === 'row' ? tableId : cellId;
+      node.setAttribute(`data-${key}`, value[key] || identityMarker());
+    });
+
+    return node;
+  }
+
+  static formats(domNode) {
+    return CELL_IDENTITY_KEYS.reduce((formats, attribute) => {
+      if (domNode.hasAttribute(`data-${attribute}`)) {
```
Comment (ksercs) [inline formats/table/index.js:23]: let's move `data-${attribute}` to a variable?

File: formats/table/index.js [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+      this.wrap(this.statics.requiredContainer.blotName, { row: rowId });
+    }
+
+    super.optimize(...args);
+  }
+
+  format(name, value) {
+    if (CELL_IDENTITY_KEYS.indexOf(name) > -1) {
+      if (value) {
+        this.domNode.setAttribute(`data-${name}`, value);
```
Comment (ksercs) [inline formats/table/index.js:46]: let's move `data-${name}` to a variable?

File: formats/table/index.js [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+    super.optimize(...args);
+  }
+}
+BaseCell.tagName = ['TD', 'TH'];
+
+class TableCell extends BaseCell {
+  static create(value) {
+    const node = super.create();
+    const attrName = 'data-row';
+    if (value && value.row) {
```
Comment (ksercs) [inline formats/table/index.js:153]:
```suggestion
    if (value?.row) {
```

File: formats/table/index.js [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+
+  rowOffset() {
+    if (this.parent) {
+      return this.parent.children.indexOf(this);
+    }
+    return -1;
+  }
+
+  table() {
+    return this.parent && this.parent.parent;
```
Comment (ksercs) [inline formats/table/index.js:249]:
```suggestion
    return this.parent?.parent;
```

File: formats/table/index.js [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+  }
+
+  static create(value) {
+    const node = super.create(value);
+    if (value && value.row) {
```
Comment (ksercs) [inline formats/table/index.js:254]:
```suggestion
    if (value?.row) {
```

File: formats/table/index.js [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+      node.setAttribute('data-row', value.row);
+    }
+    return node;
+  }
+
+  formats() {
+    return ['row'].reduce((formats, attrName) => {
+      if (this.domNode.hasAttribute(`data-${attrName}`)) {
```
Comment (ksercs) [inline formats/table/index.js:262]: let's move `data-${attrName}` to a variable?

File: modules/clipboard.js [hunk 7]
```diff
@@ -76,6 +73,7 @@ class Clipboard extends Module {
     this.quill.root.addEventListener('cut', e => this.onCaptureCopy(e, true));
     this.quill.root.addEventListener('paste', this.onCapturePaste.bind(this));
     this.matchers = [];
+    this.tableBlots = options.tableBlots || [];
```
Comment (ksercs) [inline modules/clipboard.js:76]:
```suggestion
    this.tableBlots = options.tableBlots ?? [];
```

File: modules/table/index.js [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    this.quill.clipboard.addMatcher(ELEMENT_NODE, matchDimensions);
+  }
+
+  addKeyboardHandlers() {
+    const bindings = Table.keyboardBindings;
+    Object.keys(bindings).forEach(name => {
+      if (bindings[name]) {
+        this.quill.keyboard.addBinding(bindings[name]);
+      }
+    });
```
Comment (ksercs) [inline modules/table/index.js:59]: mb this way?
```
    Object.values(bindings).forEach(binding => {
      if (binding) {
        this.quill.keyboard.addBinding(binding);
      }
    });
```

File: modules/table/index.js [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+
+    const offset = table.offset();
+    table.remove();
+    this.quill.update(Quill.sources.USER);
+    this.quill.setSelection(offset, Quill.sources.SILENT);
+  }
+
+  getTable(range = this.quill.getSelection()) {
+    if (!isDefined(range)) {
+      return [null, null, null, -1];
```
Comment (ksercs) [inline modules/table/index.js:102]: mb move it to a constant?

File: modules/table/index.js [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+      );
+      this.quill.insertText(
+        range.index,
+        '\n',
+        lineFormats.tableCellLine,
+        Quill.sources.USER,
+      );
+      this.quill.setSelection(range.index + 1, Quill.sources.SILENT);
+      this.quill.focus();
+      Object.keys(context.format).forEach(name => {
```
Comment (ksercs) [inline modules/table/index.js:290]: let's use `Object.entries`?

File: modules/table/index.js [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+  'table cell enter': {
+    key: 'enter',
+    shiftKey: null,
+    format: ['tableCellLine', 'tableHeaderCellLine'],
+    handler(range, context) {
+      if (this.quill.selection && this.quill.selection.composing) return;
+      if (range.length > 0) {
+        this.quill.scroll.deleteAt(range.index, range.length);
+      }
+      const lineFormats = Object.keys(context.format).reduce(
```
Comment (ksercs) [inline modules/table/index.js:270]: let's use `Object.entries`?

File: modules/table/utils/make_table_arrow_handler.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    collapsed: true,
+    format: formats,
+    handler(range, context) {
+      const key = up ? 'prev' : 'next';
+      const { line } = context;
+      const cell =
+        line.statics.blotName.indexOf('Line') > -1 ? line.parent : line;
+      const targetTablePart = cell.parent.parent[key];
+      const targetRow =
+        cell.parent[key] || (targetTablePart && targetTablePart.children.head);
```
Comment (ksercs) [inline modules/table/utils/make_table_arrow_handler.js:15]:
```suggestion
        cell.parent[key] || targetTablePart?.children.head;
```

**Review by ksercs — DISMISSED** [review]

**Review by alexander-kotov-dx — DISMISSED** [review]

File: modules/table/index.js (hunk 8 above)
Comment (DokaRus) [reply modules/table/index.js:59]: IE11 :( Some of the changes we want to provide privately in 21.1

File: modules/table/index.js (hunk 10 above)
Comment (DokaRus) [reply modules/table/index.js:290]: IE11 https://github.com/DevExpress/devextreme-quill/pull/15#discussion_r713854194

File: modules/table/index.js (hunk 11 above)
Comment (DokaRus) [reply modules/table/index.js:270]: IE11 https://github.com/DevExpress/devextreme-quill/pull/15#discussion_r713854194

**Review by ksercs — APPROVED** [review]

**Review by Zedwag — APPROVED** [review]

**Review by alexanderPolosatov — APPROVED** [review]

**Pull request merged**